konnectivity-client/pkg/client/client.go

```diff
@@ -125,8 +125,12 @@ func (t *grpcTunnel) serve(c clientConn) {
 				select {
 				case ch <- result:
```

Comment:
at first glance, this looks like it's introducing a race in the opposite direction... 

we're sending from `err := t.stream.Send(req)` before we start receiving from the channel at `case res := <-resCh`

if the dial completes and we try to send the result to the channel before the receiving code reaches `case res := <-resCh`, won't we hit the default case and fail/close?

Reply:
Walter had raised a similar concern fwiw.

We _think_ that race condition is possible, although very unlikely, because `case res := <-resCh` happens immediately after `t.stream.Send(req) `. And `t.stream.Send(req) ` is a non-blocking call, in that it doesn't wait until the message is received by the other side. This dial response received here `case ch <- result` also only happens after a full round trip of the dial response from konnectivity agent -> konnectivity server -> kube-apiserver. So given the sending side of the channel requires a full roundtrip on the transport the race condition seems really unlikely. 

However, even if that race condition can happen, the change at least cleans up the gRPC client connection here, instead of hanging forever and waiting til the caller of DialContext cancels the context anyways.

Reply:
hmm, a fix that trades one race for a probably less likely race is hard for me to reason about the safety of... and certainly not one I'd want to backport to 3 existing release branches without soaking/proving it out. did we look for a way to do this in a non-racy way? https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/339 came to mind... we're already giving the serve loop a result channel to send to... can we also give it a cancel channel to unblock itself on that DialContext closes when it has given up on receiving?